code/parse/sexp.cpp

```diff
@@ -15073,6 +15075,327 @@ void sexp_toggle_asteroid_field(int n)
+	obj_delete_all_that_should_be_dead();
+
+	int num_asteroids = 0, asteroid_speed = 0;
+	for (int i = 0; i < 2; i++) {
+
+		int temp = eval_num(n, is_nan, is_nan_forever);
+		if (is_nan || is_nan_forever)
+			return;
+
+		if (i == 0)
+			num_asteroids = temp;
+		else
+			asteroid_speed = temp;
+		n = CDR(n);
+	}
+
+	int debris1 = -1;
+	if (n >= 0) {
+		debris1 = get_asteroid_position(CTEXT(n));
+		n = CDR(n);
+	}
+
+	int debris2 = -1;
+	if (n >= 0) {
+		debris2 = get_asteroid_position(CTEXT(n));
+		n = CDR(n);
+	}
+
+	int debris3 = -1;
+	if (n >= 0) {
+		debris3 = get_asteroid_position(CTEXT(n));
+		n = CDR(n);
+	}
+
+	if ((debris1 == -1) && (debris2 == -1) && (debris3 == -1))
+		return;
+
+	int o_minx = -1000, o_miny = -1000, o_minz = -1000;
+	int o_maxx = 1000, o_maxy = 1000, o_maxz = 1000;
+	if (n >= 0) {
+		for (int i = 0; i < 6; i++) {
+
+			int temp = eval_num(n, is_nan, is_nan_forever);
+			if (is_nan || is_nan_forever)
+				return;
+
+			if (i == 0)
+				o_minx = temp;
+			else if (i == 1)
+				o_maxx = temp;
+			else if (i == 2)
+				o_miny = temp;
+			else if (i == 3)
+				o_maxy = temp;
+			else if (i == 4)
+				o_minz = temp;
+			else
+				o_maxz = temp;
+			n = CDR(n);
+		}
+	}
+
+	SCP_vector<SCP_string> targets;
+	if (n >= 0) {
+		int i = 0;
+		for (; n >= 0; true ? n = CDR(n) : n = -1) {
+			auto ship_entry = eval_ship(n);
+			if (!ship_entry)
+				continue;
+
+			targets[i] = ship_entry->name;
+			i++;
+		}
+	}
+
+	if (num_asteroids == 0) {
+		return;
+	} else if (num_asteroids > MAX_ASTEROIDS) {
+		num_asteroids = MAX_ASTEROIDS;
+	}
+	Asteroid_field.num_initial_asteroids = num_asteroids;
+
+	Asteroid_field.field_type = FT_PASSIVE;
+
+	vm_vec_rand_vec_quick(&Asteroid_field.vel);
+	vm_vec_scale(&Asteroid_field.vel, (float)asteroid_speed);
+	Asteroid_field.speed = (float)asteroid_speed;
+	Asteroid_field.debris_genre = DG_SHIP;
+
+	Asteroid_field.field_debris_type[0] = -1;
+	Asteroid_field.field_debris_type[1] = -1;
+	Asteroid_field.field_debris_type[2] = -1;
+
+	int count = 0;
+	for (int i = 0; i < MAX_ACTIVE_DEBRIS_TYPES; i++) {
+		if (debris1 >= 0) {
+			Asteroid_field.field_debris_type[i] = debris1;
+			debris1 = -1;
+			count++;
+			continue;
+		}
+		if (debris2 >= 0) {
+			Asteroid_field.field_debris_type[i] = debris2;
+			debris2 = -1;
+			count++;
+			continue;
+		}
+		if (debris3 >= 0) {
+			Asteroid_field.field_debris_type[i] = debris3;
+			debris3 = -1;
+			count++;
+			continue;
+		}
+	}
+
+	Asteroid_field.num_used_field_debris_types = count;
+
+	Asteroid_field.min_bound = vm_vec_new((float)o_minx, (float)o_miny, (float)o_minz);
+	Asteroid_field.max_bound = vm_vec_new((float)o_maxx, (float)o_maxy, (float)o_maxz);
+
+	vec3d a_rad;
+	vm_vec_sub(&a_rad, &Asteroid_field.max_bound, &Asteroid_field.min_bound);
+	vm_vec_scale(&a_rad, 0.5f);
+	float b_rad = vm_vec_mag(&a_rad);
+
+	Asteroid_field.bound_rad = MAX(3000.0f, b_rad);
+
+	asteroid_create_all();
```

Comment:
Relatively minor, but both these SEXPs are quite 'big'. It might be better to decouple the SEXP handling and the the actual asteroid/debris handling, so the actual deletion/making of the fields are handled in functions in their respective files, which these SEXPs call, and only worry about retrieving the right arguments for it. Also makes it easier to reuse this logic in different contexts in the future.

Reply:
Good point. Taken care of.